lib/auth.ml

```diff
@@ -71,9 +66,23 @@ let sign name alg key session_id service =
   let data = to_hash name alg (Hostkey.pub_of_priv key) session_id service in
   Hostkey.sign alg key data
 
-let by_pubkey name alg pubkey session_id service signed db =
-  match lookup_user_key name pubkey db with
-  | None -> false
-  | Some pubkey ->
-    let unsigned = to_hash name alg pubkey session_id service in
-    Hostkey.verify alg pubkey ~unsigned ~signed
+let verify_signature name alg pubkey session_id service signed =
+  let unsigned = to_hash name alg pubkey session_id service in
+  Hostkey.verify alg pubkey ~unsigned ~signed
+
+let verify_pubkeyauth ~user { pubkey; session_id; service ; sig_alg ; signed } =
+  verify_signature user sig_alg pubkey session_id service signed
+
+let verify db user userauth =
+  match lookup_user user db, userauth with
+  | None, Pubkey pubkeyauth ->
+    verify_pubkeyauth ~user pubkeyauth && false
+  | (None | Some { password = None; _ }), Password _ -> false
+  | Some u, Pubkey pubkeyauth ->
+    verify_pubkeyauth ~user pubkeyauth &&
+    List.exists (fun pubkey -> Hostkey.pub_eq pubkey pubkeyauth.pubkey) u.keys
+  | Some { password = Some password; _ }, Password password' ->
+      let open Digestif.SHA256 in
+      let a = digest_string password
+      and b = digest_string password' in
```

Comment:
seeing it like this, I was wondering whether we should store the password as a hash or somehow derived in the memory of the server -- but this can obviously be another PR since AFAICT this PR doesn't change it, only makes it more obvious